rs/messaging/src/routing/stream_handler/tests.rs

```diff
@@ -689,1112 +539,1230 @@ fn induct_loopback_stream_with_memory_limit_impl(
+    metrics.assert_inducted_xnet_messages_eq(&[
+        (
+            LABEL_VALUE_TYPE_RESPONSE,
+            LABEL_VALUE_SENDER_SUBNET_MISMATCH,
+            1,
+        ),
+        (
+            LABEL_VALUE_TYPE_REQUEST,
+            LABEL_VALUE_SENDER_SUBNET_MISMATCH,
+            1,
+        ),
+        (LABEL_VALUE_TYPE_RESPONSE, LABEL_VALUE_SUCCESS, 1),
+    ]);
+    metrics.assert_eq_critical_errors(CriticalErrorCounts {
+        sender_subnet_mismatch: 2,
+        ..CriticalErrorCounts::default()
+    });
+}
+
+/// Common implementation for tests checking reject responses generated locally by the
+/// `StreamHandler` directly.
+fn check_stream_handler_locally_generated_reject_response_impl(
+    reason: RejectReason,
+    expected_reject_code: RejectCode,
+    expected_message: String,
+) {
+    // An outgoing stream with one request @21 in it.
+    let outgoing_stream_config = StreamConfigX {
+        begin: 21,
+        messages: vec![*TO_REMOTE_REQUEST_CONFIG],
+        ..StreamConfigX::default()
+    };
+
+    // An incoming stream slice with a reject signal for the request @21 in it.
+    let incoming_stream_slice_config = StreamSliceConfigX {
+        signals_end: 22,
+        reject_signals: vec![RejectSignal::new(reason, 21.into())],
+        ..StreamSliceConfigX::default()
+    };
+
+    let (stream_handler, state, slices, _) = with_test_setup(
+        btreemap![REMOTE_SUBNET => outgoing_stream_config],
+        btreemap![REMOTE_SUBNET => incoming_stream_slice_config],
+    );
+
+    // Generate the expected reject response for the request in `outgoing_stream`.
+    let request = request_ref_in_stream(state.get_stream(&REMOTE_SUBNET), 21);
+    let reject_response = Response {
+        originator: request.sender,
+        respondent: request.receiver,
+        originator_reply_callback: request.sender_reply_callback,
+        refund: request.payment,
+        response_payload: Payload::Reject(RejectContext::new_with_message_length_limit(
+            expected_reject_code,
+            expected_message,
+            MR_SYNTHETIC_REJECT_MESSAGE_MAX_LEN,
+        )),
+        deadline: request.deadline,
+    };
+
+    // The expected state has this reject response inducted.
+    let mut expected_state = state.clone();
+    push_input(&mut expected_state, reject_response.into()).unwrap();
+
+    // The expected stream is gc'ed.
+    let expected_stream = stream_from_config(StreamConfigX {
+        begin: 22,
+        ..StreamConfigX::default()
     });
+    expected_state.with_streams(btreemap![REMOTE_SUBNET => expected_stream]);
+
+    let inducted_state = stream_handler.process_stream_slices(state, slices);
+
+    assert_eq!(inducted_state, expected_state);
 }
 
 #[test]
-fn reroute_rejected_response_success() {
-    with_test_replica_logger(|log| {
-        // Empty initial state.
-        let (stream_handler, mut initial_state, _) = new_fixture(&log);
```

Comment:
This test is redundant since what it tests is already included in `garbage_collect_local_state_with_reject_signals_for_response_success()`. It is the only test that tests one of the 'internal functions' of the stream handler directly. It therefore also doesn't quite fit with my test framwork very well.